Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -237,6 +245,189 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
         parseWebAccessibleResourcesVersion2();
 }
 
+static int toAPI(WebExtension::Error error)
+{
+    switch (error) {
+    case WebExtension::Error::Unknown:
+        return static_cast<int>(WebExtension::APIError::Unknown);
+    case WebExtension::Error::ResourceNotFound:
+        return static_cast<int>(WebExtension::APIError::ResourceNotFound);
+    case WebExtension::Error::InvalidManifest:
+        return static_cast<int>(WebExtension::APIError::InvalidManifest);
+    case WebExtension::Error::UnsupportedManifestVersion:
+        return static_cast<int>(WebExtension::APIError::UnsupportedManifestVersion);
+    case WebExtension::Error::InvalidDeclarativeNetRequest:
+        return static_cast<int>(WebExtension::APIError::InvalidDeclarativeNetRequestEntry);
+    case WebExtension::Error::InvalidBackgroundPersistence:
+        return static_cast<int>(WebExtension::APIError::InvalidBackgroundPersistence);
+    case WebExtension::Error::InvalidResourceCodeSignature:
+        return static_cast<int>(WebExtension::APIError::InvalidResourceCodeSignature);
+    default:
+        return static_cast<int>(WebExtension::APIError::InvalidManifestEntry);
+    }
+}
+
+Ref<API::Error> WebExtension::createError(Error error, const String& customLocalizedDescription, RefPtr<API::Error> underlyingError)
+{
+    auto errorCode = toAPI(error);
+    String localizedDescription;
+    RefPtr<JSON::Value> actionManifest;
+
+    if (supportsManifestVersion(3))
+        actionManifest = manifestObject()->getValue(actionManifestKey);
+    else {
+        actionManifest = manifestObject()->getValue(browserActionManifestKey);
+        if (!actionManifest)
+            actionManifest = manifestObject()->getValue(pageActionManifestKey);
+    }
```

Comment:
This should move since `case Error::InvalidActionIcon: { … }` since it is only needed there.